dot/node_test.go

```diff
@@ -0,0 +1,400 @@
+		basepath string
+	}
+	tests := []struct {
+		name         string
+		args         args
+		wantNodename string
+		err          error
+	}{
+		{
+			name:         "working example",
+			args:         args{basepath: basePath},
+			wantNodename: "nodeName",
+		},
+		{
+			name: "wrong basepath test",
+			args: args{basepath: "wrong_path"},
+			err:  errors.New("Key not found"),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gotNodename, err := LoadGlobalNodeName(tt.args.basepath)
+			if tt.err != nil {
+				assert.EqualError(t, err, tt.err.Error())
+				err := os.RemoveAll(tt.args.basepath)
+				require.NoError(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+
+			assert.Equal(t, tt.wantNodename, gotNodename)
+		})
+	}
+}
+
+func TestNewNode(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	mockTelemetryClient := NewMockClient(ctrl)
+	mockTelemetryClient.EXPECT().SendMessage(gomock.Any())
+
+	initConfig := NewTestConfig(t)
+
+	genFile := newTestGenesisRawFile(t, initConfig)
+
+	networkConfig := &network.Config{
+		BasePath:    t.TempDir(),
+		NoBootstrap: true,
+		NoMDNS:      true,
+	}
+	testNetwork := createTestService(t, networkConfig)
+
+	config := state.Config{
+		Path:     initConfig.Global.BasePath,
+		LogLevel: initConfig.Log.StateLvl,
+	}
+
+	dotConfig := &Config{
+		Global:  GlobalConfig{BasePath: initConfig.Global.BasePath},
+		Init:    InitConfig{Genesis: genFile},
+		Account: AccountConfig{Key: "alice"},
+		Core: CoreConfig{
+			Roles:           types.FullNodeRole,
+			WasmInterpreter: wasmer.Name,
+		},
+	}
+
+	dotConfig.Init = InitConfig{Genesis: genFile}
+	dotConfig.Account = AccountConfig{Key: "alice"}
+	dotConfig.Core.Roles = types.FullNodeRole
+	dotConfig.Core.WasmInterpreter = wasmer.Name
+
+	ks, err := initKeystore(dotConfig)
+	assert.NoError(t, err)
+	m := NewMocknodeBuilderIface(ctrl)
+	m.EXPECT().nodeInitialised(dotConfig.Global.BasePath).Return(true)
+	m.EXPECT().createStateService(dotConfig).DoAndReturn(func(cfg *Config) (*state.Service, error) {
+		stateSrvc := state.NewService(config)
+		// create genesis from configuration file
+		gen, err := genesis.NewGenesisFromJSONRaw(cfg.Init.Genesis)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load genesis from file: %w", err)
+		}
+		// create trie from genesis
+		trie, err := genesis.NewTrieFromGenesis(gen)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create trie from genesis: %w", err)
+		}
+		// create genesis block from trie
+		header, err := genesis.NewGenesisBlockFromTrie(trie)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create genesis block from trie: %w", err)
+		}
+		stateSrvc.Telemetry = mockTelemetryClient
+		err = stateSrvc.Initialise(gen, header, trie)
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialise state service: %s", err)
+		}
+
+		err = stateSrvc.SetupBase()
+		if err != nil {
+			return nil, fmt.Errorf("cannot setup base: %w", err)
+		}
+		return stateSrvc, nil
```

Comment:
Yes, I think once I've make stateService an interface and mocked those function, then I can mock these.

Reply:
Why not just return a dull service like `&state.Service{}`?

Reply:
When I test with null service `&state.Service{}` the test fails with nil pointer error, because in newNode `stateSrvc.Base.LoadGenesisData()` is called.  It may be overkill to init state.Service, which is why were going to make an interface for state.Service so required methods can be mocked.